File: lib/src/widgets/wavbuf.dart
// ignore_for_file: curly_braces_in_flow_control_structures

import 'dart:ffi';
import 'dart:typed_data';
import 'dart:math';

class WaveBuf {
  final Uint8List _buf= Uint8List(48000*2*4);
  int _cur=0;
  bool push(Uint8List v) {
    if (_cur+v.length>_buf.length)
      return false;
    for (int i=0; i<v.length; i++, _cur++) {
      _buf[_cur]= v[i];
    }
    return true;
  }
  int pop(int n, {Uint8List? dst}) {
    n= min(n, _cur);
    if (dst!=null) {
      for (int i=0; i<n; i++)
        dst[i] = _buf[i];
    }
    int r= _cur-n;
    for (int i=0; i<r; i++)
      _buf[i]= _buf[n+i];
    _cur= r;
    return n;
  }
  int read(int n, Pointer<Uint8> p) {
    n= min(n, _cur);
    for (int i=0; i<n; i++)
      p[i]= _buf[i];
    return n;
  }
  int copy(int n, Uint8List dst) {
    n= min(n, _cur);
    for (int i=0; i<n; i++)
      dst[i]= _buf[i];
    return n;
  }
  int get length => _cur;
  void clear() => _cur=0;
}

File: ios/watch_ios Watch App/wavbuf.swift

//
//  wbuf.swift
//  Runner
//
//  Created by 방경식 on 5/29/25.
//

import Foundation

class WaveBuf {
    private var buf: [UInt8]
    private var cur: Int = 0
    private let capacity: Int

    init(size: Int = 48000 * 2) {
        self.capacity = size
        self.buf = [UInt8](repeating: 0, count: size)
    }

    /// 데이터 삽입
    func push(_ data: [UInt8]) -> Bool {
        if cur + data.count > capacity {
            return false
        }
        for byte in data {
            buf[cur] = byte
            cur += 1
        }
        return true
    }

    /// 앞에서 n개 제거 (선택적으로 dst로 복사)
    @discardableResult
    func pop(_ n: Int, dst: UnsafeMutablePointer<UInt8>? = nil) -> Int {
        let len = min(n, cur)
        if let d = dst {
            for i in 0..<len {
                d[i] = buf[i]
            }
        }

        let remaining = cur - len
        for i in 0..<remaining {
            buf[i] = buf[len + i]
        }

        cur = remaining
        return len
    }

    /// Pointer로 복사 (C 코드와 연동할 때)
    @discardableResult
    func read(_ n: Int, to ptr: UnsafeMutablePointer<UInt8>) -> Int {
        let len = min(n, cur)
        for i in 0..<len {
            ptr[i] = buf[i]
        }
        return len
    }

    /// Swift 배열로 복사
    @discardableResult
    func copy(_ n: Int, to dst: inout [UInt8]) -> Int {
        let len = min(n, cur)
        for i in 0..<len {
            dst[i] = buf[i]
        }
        return len
    }

    var length: Int {
        return cur
    }

    func clear() {
        cur = 0
    }
}
